File: Examples/counts_matrix_projection.jl
using BioToolkit
using CSV
using DataFrames
using Statistics
using LinearAlgebra
using SparseArrays
using Plots

function load_counts_matrix(csv_path::AbstractString)
    counts_df = CSV.read(csv_path, DataFrame)
    counts_matrix = Matrix{Int}(counts_df[:, 2:end])
    return sparse(counts_matrix)
end

function procrustes_julia(source::AbstractMatrix{<:Real}, target::AbstractMatrix{<:Real})
    source_matrix = Matrix{Float64}(source)
    target_matrix = Matrix{Float64}(target)

    source_mean = vec(mean(source_matrix, dims=1))
    target_mean = vec(mean(target_matrix, dims=1))
    source_centered = source_matrix .- reshape(source_mean, 1, :)
    target_centered = target_matrix .- reshape(target_mean, 1, :)

    factorization = svd(transpose(source_centered) * target_centered)
    rotation = factorization.U * factorization.Vt
    scale = sum(factorization.S) / max(sum(abs2, source_centered), eps(Float64))
    return scale, rotation, source_mean, target_mean
end

function counts_matrix_projection_report(; csv_path::AbstractString=joinpath(@__DIR__, "counts_matrix.csv"), reference_cells::Int=400, n_components::Int=30, save_plot::Bool=true, plot_path::AbstractString=joinpath(@__DIR__, "counts_matrix_projection.png"))
    counts_sparse = load_counts_matrix(csv_path)
    n_genes, n_cells = size(counts_sparse)
    reference_cells < n_cells || throw(ArgumentError("reference_cells must be smaller than the number of cells"))

    gene_ids = ["gene_$(i)" for i in 1:n_genes]
    cell_ids = ["cell_$(i)" for i in 1:n_cells]
    reference_idx = 1:reference_cells
    query_idx = (reference_cells + 1):n_cells

    sce_full = SingleCellExperiment(counts_sparse, gene_ids, cell_ids)
    full_normalized = normalize_counts(sce_full)
    truth_coords = run_pca(sce_full; normalized=full_normalized, n_components=n_components, use_variable_features=false)

    sce_reference = SingleCellExperiment(counts_sparse[:, reference_idx], gene_ids, cell_ids[reference_idx])
    sce_query = SingleCellExperiment(counts_sparse[:, query_idx], gene_ids, cell_ids[query_idx])

    projection_model = fit_singlecell_projection_model(
        sce_reference;
        n_components=n_components,
        use_variable_features=false,
        scale_features=false,
    )
    projected_reference = project_singlecell(sce_reference, projection_model; reduction_name="counts_matrix_projection_reference")
    projected_query = project_singlecell(sce_query, projection_model; reduction_name="counts_matrix_projection_query")

    true_reference = truth_coords[reference_idx, :]
    true_query = truth_coords[query_idx, :]

    scale, rotation, projected_mean, truth_mean = procrustes_julia(projected_reference, true_reference)
    aligned_query = scale .* ((Matrix{Float64}(projected_query) .- reshape(projected_mean, 1, :)) * rotation) .+ reshape(truth_mean, 1, :)

    pc1_corr = cor(aligned_query[:, 1], true_query[:, 1])
    rmse = sqrt(mean((aligned_query .- true_query) .^ 2))

    println("------------------------------------------------")
    println("COUNTS MATRIX PROJECTION REPORT")
    println("------------------------------------------------")
    println("Genes: $(n_genes)")
    println("Cells: $(n_cells)")
    println("Reference cells: $(reference_cells)")
    println("Query cells: $(length(query_idx))")
    println("Julia RMSE (on R data): $(round(rmse, digits=4))")
    println("Julia Cor PC1: $(round(pc1_corr, digits=4))")

    if save_plot
        p = scatter(
            true_query[:, 1],
            true_query[:, 2];
            label="True Query",
            marker=:circle,
            alpha=0.75,
            title="Counts Matrix Projection",
            xlabel="PC1",
            ylabel="PC2",
        )
        scatter!(
            p,
            aligned_query[:, 1],
            aligned_query[:, 2];
            label="Projected Query",
            marker=:cross,
            alpha=0.75,
        )
        savefig(p, plot_path)
        println("Plot saved to $(plot_path)")
        println("------------------------------------------------")
    end

    return (
        rmse=rmse,
        pc1_corr=pc1_corr,
        projected_query=aligned_query,
        true_query=true_query,
    )
end

if abspath(PROGRAM_FILE) == @__FILE__
    report = counts_matrix_projection_report()
    @assert report.pc1_corr > 0.99 "Projection consistency failed: PC1 correlation too low"
    @assert report.rmse < 1.0 "Projection consistency failed: RMSE too high"
    println("SUCCESS: counts_matrix.csv projection benchmark passed.")
end

File: Examples/counts_matrix_projection.r
# Install packages
if (!require("Matrix")) install.packages("Matrix")
if (!require("ggplot2")) install.packages("ggplot2")
library(Matrix)
library(ggplot2)

set.seed(42)

# 1. Generate Data
n_genes <- 1000
n_cells <- 500
latent_dim <- 4

# (Your generation logic - simplified for reproducibility)
gene_loadings <- matrix(0, nrow=n_genes, ncol=latent_dim)
gene_groups <- sample(1:latent_dim, n_genes, replace=TRUE)
for (i in 1:n_genes) {
  gene_loadings[i, gene_groups[i]] <- 2.5 + runif(1)
  gene_loadings[i, -gene_groups[i]] <- 0.1 * rnorm(latent_dim-1)
}
cell_scores <- matrix(rnorm(latent_dim * n_cells), nrow=latent_dim)
signal <- gene_loadings %*% cell_scores / sqrt(latent_dim)
baseline <- 35.0 + 4.0 * rnorm(n_genes)
noisy_counts <- round(baseline + 14.0 * signal + 0.35 * matrix(rnorm(n_genes * n_cells), nrow=n_genes))
noisy_counts[noisy_counts < 2] <- 0
counts <- Matrix(noisy_counts, sparse=TRUE)

# 2. Save Data for Julia
write.csv(as.matrix(counts), "counts_matrix.csv")
cat("Saved counts_matrix.csv for Julia.\n")

# 3. Run R Benchmark
normalize_counts <- function(counts) {
  cs <- colSums(counts)
  cs[cs==0] <- 1
  log1p(t(t(counts) / cs) * 10000)
}

procrustes_align <- function(source, target) {
  sm <- colMeans(source); tm <- colMeans(target)
  s_c <- sweep(source, 2, sm); t_c <- sweep(target, 2, tm)
  svd_res <- svd(crossprod(s_c, t_c))
  rot <- svd_res$u %*% t(svd_res$v)
  sc <- sum(svd_res$d) / max(sum(s_c^2), .Machine$double.eps)
  list(aligned=sc * (s_c %*% rot) + matrix(tm, nrow=nrow(source), ncol=ncol(source), byrow=T), 
       rot=rot, scale=sc, sm=sm, tm=tm)
}

n_comp <- 30
ref_idx <- 1:400
query_idx <- 401:500

full_norm <- normalize_counts(counts)
full_pca <- prcomp(t(full_norm), center=T, scale.=F, rank.=n_comp)
full_truth <- full_pca$x

counts_ref <- counts[, ref_idx]
counts_query <- counts[, query_idx]

ref_norm <- normalize_counts(counts_ref)
ref_pca <- prcomp(t(ref_norm), center=T, scale.=F, rank.=n_comp)

# Projection
proj_query <- scale(t(normalize_counts(counts_query)), ref_pca$center, F) %*% ref_pca$rotation
proj_ref <- scale(t(ref_norm), ref_pca$center, F) %*% ref_pca$rotation

# Alignment
align_res <- procrustes_align(proj_ref, full_truth[ref_idx, ])
aligned_query <- align_res$scale * (sweep(proj_query, 2, align_res$sm) %*% align_res$rot) + matrix(align_res$tm, nrow=100, ncol=30, byrow=T)

# Report
cat("R RMSE:", sqrt(mean((aligned_query - full_truth[query_idx, ])^2)), "\n")
cat("R Cor PC1:", cor(aligned_query[,1], full_truth[query_idx, 1]), "\n")


# Output -->
# Saved counts_matrix.csv for Julia.
# R RMSE: 0.1181703 
# R Cor PC1: 1 